File: mobile/lib/features/products/models/product.dart
class QaleenSize {
  final double height;
  final double width;
  final int pieces;

  const QaleenSize({required this.height, required this.width, required this.pieces});

  factory QaleenSize.fromJson(Map<String, dynamic> json) => QaleenSize(
        height: (json['height'] as num?)?.toDouble() ?? 0,
        width: (json['width'] as num?)?.toDouble() ?? 0,
        pieces: (json['pieces'] as num?)?.toInt() ?? 0,
      );

  Map<String, dynamic> toJson() => {'height': height, 'width': width, 'pieces': pieces};
}

class Product {
  final String id;
  final String name;
  final String sku;
  final String barcode;
  final String? categoryId;
  final String? categoryName;
  final String brand;
  final String supplier;
  final String productType;
  final double carpetWidth;
  final double carpetHeight;
  final int carpetPieces;
  final double costPerSqft;
  final double costPerPiece;
  final List<QaleenSize> qaleenSizes;
  final double meterLength;
  final double costPerMeter;
  final double costPrice;
  final double sellingPrice;
  final int quantity;
  final int lowStockThreshold;
  final String color;
  final String size;
  final String description;
  final List<String> images;
  final String? shopId;
  final String? shopName;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Product({
    required this.id,
    required this.name,
    this.sku = '',
    this.barcode = '',
    this.categoryId,
    this.categoryName,
    this.brand = '',
    this.supplier = '',
    this.productType = 'qaleen',
    this.carpetWidth = 0,
    this.carpetHeight = 0,
    this.carpetPieces = 0,
    this.costPerSqft = 0,
    this.costPerPiece = 0,
    this.qaleenSizes = const [],
    this.meterLength = 0,
    this.costPerMeter = 0,
    required this.costPrice,
    this.sellingPrice = 0,
    this.quantity = 0,
    this.lowStockThreshold = 5,
    this.color = '',
    this.size = '',
    this.description = '',
    this.images = const [],
    this.shopId,
    this.shopName,
    this.createdAt,
    this.updatedAt,
  });

  bool get isLowStock => quantity <= lowStockThreshold;

  double get stockValue => quantity * costPrice;

  factory Product.fromJson(Map<String, dynamic> json) {
    final category = json['category'];
    String? catId;
    String? catName;
    if (category is Map<String, dynamic>) {
      catId = (category['id'] ?? category['_id'])?.toString();
      catName = category['name']?.toString();
    } else if (category != null) {
      catId = category.toString();
    }
    final shop = json['shop'];
    String? sId;
    String? sName;
    if (shop is Map<String, dynamic>) {
      sId = (shop['id'] ?? shop['_id'])?.toString();
      sName = shop['name']?.toString();
    } else if (shop != null) {
      sId = shop.toString();
    }

    return Product(
      id: (json['id'] ?? json['_id']).toString(),
      name: json['name']?.toString() ?? '',
      sku: json['sku']?.toString() ?? '',
      barcode: json['barcode']?.toString() ?? '',
      categoryId: catId,
      categoryName: catName,
      brand: json['brand']?.toString() ?? '',
      supplier: json['supplier']?.toString() ?? '',
      productType: json['productType']?.toString() ?? 'qaleen',
      carpetWidth: (json['carpetWidth'] as num?)?.toDouble() ?? 0,
      carpetHeight: (json['carpetHeight'] as num?)?.toDouble() ?? 0,
      carpetPieces: (json['carpetPieces'] as num?)?.toInt() ?? 0,
      costPerSqft: (json['costPerSqft'] as num?)?.toDouble() ?? 0,
      costPerPiece: (json['costPerPiece'] as num?)?.toDouble() ?? 0,
      qaleenSizes: (json['qaleenSizes'] as List?)
              ?.map((e) => QaleenSize.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      meterLength: (json['meterLength'] as num?)?.toDouble() ?? 0,
      costPerMeter: (json['costPerMeter'] as num?)?.toDouble() ?? 0,
      costPrice: (json['costPrice'] as num?)?.toDouble() ?? 0,
      sellingPrice: (json['sellingPrice'] as num?)?.toDouble() ?? 0,
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
      lowStockThreshold: (json['lowStockThreshold'] as num?)?.toInt() ?? 5,
      color: json['color']?.toString() ?? '',
      size: json['size']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      images: (json['images'] as List?)?.map((e) => e.toString()).toList() ?? const [],
      shopId: sId,
      shopName: sName,
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? ''),
      updatedAt: DateTime.tryParse(json['updatedAt']?.toString() ?? ''),
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'sku': sku,
        'barcode': barcode,
        'category': categoryId,
        'brand': brand,
        'supplier': supplier,
        'productType': productType,
        'carpetWidth': carpetWidth,
        'carpetHeight': carpetHeight,
        'carpetPieces': carpetPieces,
        'costPerSqft': costPerSqft,
        'costPerPiece': costPerPiece,
        'qaleenSizes': qaleenSizes.map((s) => s.toJson()).toList(),
        'meterLength': meterLength,
        'costPerMeter': costPerMeter,
        'costPrice': costPrice,
        'sellingPrice': sellingPrice,
        'lowStockThreshold': lowStockThreshold,
        'color': color,
        'size': size,
        'description': description,
        'images': images,
      };
}

File: backend/src/models/Product.js
const mongoose = require('mongoose');

const qaleenSizeSchema = new mongoose.Schema(
  { height: { type: Number, required: true }, width: { type: Number, required: true }, pieces: { type: Number, required: true, min: 1 } },
  { _id: false }
);

const carpetPieceSchema = new mongoose.Schema(
  {
    width: { type: Number, required: true },
    height: { type: Number, required: true },
    area: { type: Number, default: 0 },
    color: { type: String, default: '' },
    image: { type: String, default: '' },
  },
  { _id: false }
);

const productSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true },
    sku: { type: String, default: '', trim: true },
    barcode: { type: String, default: '', trim: true },
    category: { type: mongoose.Schema.Types.ObjectId, ref: 'Category', default: null },
    brand: { type: String, default: '' },
    supplier: { type: String, default: '' },
    productType: { type: String, enum: ['carpet', 'qaleen', 'meter'], default: 'qaleen' },

    carpetWidth: { type: Number, default: 0 },
    carpetHeight: { type: Number, default: 0 },
    carpetPieces: { type: Number, default: 0 },
    carpetPiecesData: { type: [carpetPieceSchema], default: [] },
    costPerSqft: { type: Number, default: 0 },

    costPerPiece: { type: Number, default: 0 },
    qaleenSizes: { type: [qaleenSizeSchema], default: [] },

    meterLength: { type: Number, default: 0 },
    costPerMeter: { type: Number, default: 0 },

    costPrice: { type: Number, required: true, min: 0, default: 0 },
    sellingPrice: { type: Number, min: 0, default: 0 },
    quantity: { type: Number, required: true, min: 0, default: 0 },
    lowStockThreshold: { type: Number, default: 5 },
    color: { type: String, default: '' },
    size: { type: String, default: '' },
    description: { type: String, default: '' },
    images: { type: [String], default: [] },
    shop: { type: mongoose.Schema.Types.ObjectId, ref: 'Shop', required: true },
    isDeleted: { type: Boolean, default: false },
    deletedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    deletedAt: { type: Date, default: null },
    deleteReason: { type: String, default: '' },
  },
  { timestamps: true }
);

productSchema.index({ shop: 1, isDeleted: 1 });
productSchema.index({ name: 'text', sku: 'text', barcode: 'text' });

module.exports = mongoose.model('Product', productSchema);
